File: src/main/java/com/codewars/kyu6/FIXMEHello.java
package com.codewars.kyu6;

import java.util.LinkedHashMap;

/**
 * @author Zurbaevi Nika
 */
class FIXMEHello extends LinkedHashMap<String, String> {
    FIXMEHello setAge(int age) {
        put("age", " I am " + age + ".");
        return this;
    }

    FIXMEHello setSex(char sex) {
        put("sex", sex == 'M' ? " I am male." : " I am female.");
        return this;
    }

    FIXMEHello setName(String name) {
        put("name", " My name is " + name + ".");
        return this;
    }

    String hello() {
        return "Hello." + String.join("", values());
    }
}

File: src/main/kotlin/com/codewars/kyu6/FIXMEHello.kt
package com.codewars.kyu6

/**
 * @author Zurbaevi Nika
 */
class Dinglemouse {

    enum class OrderItem {
        Name,
        Age,
        Sex
    }

    private var name: String = ""
    private var age: Int = 0
    private var sex: Char = ' '

    private val order = mutableListOf<OrderItem>()

    fun setAge(age: Int): Dinglemouse {
        this.age = age
        addToOrder(OrderItem.Age)
        return this
    }

    fun setSex(sex: Char): Dinglemouse {
        this.sex = sex
        addToOrder(OrderItem.Sex)
        return this
    }

    fun setName(name: String): Dinglemouse {
        this.name = name
        addToOrder(OrderItem.Name)
        return this
    }

    private fun addToOrder(orderItem: OrderItem) {
        if (!order.contains(orderItem)) {
            order.add(orderItem)
        }
    }

    fun hello(): String {
        val output = StringBuilder("Hello. ")
        order.forEach {
            when (it) {
                OrderItem.Age -> output.append("I am ${age}. ")
                OrderItem.Sex -> output.append("I am ${if (sex == 'M') "male" else "female"}. ")
                OrderItem.Name -> output.append("My name is $name. ")
            }
        }
        return output.toString().trim()
    }

}